evaluation/evaluation.md

```diff
@@ -1764,3 +1768,328 @@ display_merged_filtered_results(counts, titles, total_results, combinations, lab
 </table>
 </div>
 
+
+## Addon counts
+
+We want to train an ensemble model using the individual recommenders that we already have. To optimize this ensemble model, we need some training data. We have information about what addons different users have installed, so it would make sense to check if our ensemble model would also recommend these addons to the respective users.
+
+However, there is one fundamental conflict here: To be able to make recommendations, the collaborative recommender already needs some information about which addons a user has installed. Thus, we can only use a subset of a user's installed addons for evaluation. These addons are masked and the collaborative recommender then only uses the unmasked addons. The open question is how large this subset of masked addons should be.
+
+Choosing this size is a trade-off between three factors:
+1. By masking more addons, we give the evaluation function more data to work with
+2. If we mask fewer addons, the collaborative filter can make better recommendations
+3. There are not that many users who have many addons installed. This means we need to be careful not to make our evaluation set too biased. For example, if we always mask at least five addons, then only users with more than six addons could be part of the evaluation set, which is a small subset of the entire population
+
+To be able to make this decision, it's helpful to look at the distribution of the number of addons that users have installed.
+
+
+```python
+import matplotlib.pyplot as plt
+import seaborn as sns
+%matplotlib inline
+sns.set(style="darkgrid")
+```
+
+
+```python
+addon_counts = rdd_completed\
+    .map(lambda client: len(client['installed_addons']))\
+    .map(lambda x: (x, 1))\
+    .reduceByKey(add)\
+    .collect()
+    
+addon_counts = sorted(addon_counts, key=itemgetter(0))
+```
+
+
+```python
+def addon_counts_to_proportions(addon_counts):
+    num_addons, num_addons_count = zip(*addon_counts)
+    
+    num_addons_total = float(sum(num_addons_count))
+    num_addons_count = np.array(num_addons_count) / num_addons_total
+    
+    return num_addons, num_addons_count
+```
+
+
+```python
+def plot_addon_distribution(addon_counts):
+    num_addons, num_addons_count = addon_counts_to_proportions(addon_counts)
+    
+    plt.bar(num_addons, num_addons_count, width=1.)
+    plt.title("Number of addons per user")
+    plt.xlabel("Number of addons")
+    plt.ylabel("Proportion of users")
+    plt.show()
+    
+    plt.plot(np.cumsum(num_addons_count))
```

Comment:
Lets truncate the x axis at 20-30 ish and get a better resolution on the x-axis ticks (with a grid) to make this more clear.

Reply:
The last plot in the notebook has the x-axis truncated at 15 or 20. Is this already what you're looking for?